proxy.go

```diff
@@ -177,6 +182,45 @@ func printAgentLogs(sock string) error {
 	return nil
 }
 
+func setupNotifier() chan os.Signal {
+	sigCh := make(chan os.Signal)
+	signal.Notify(sigCh, termSignal)
+
+	return sigCh
+}
+
+// Blocking function waiting for a SIGTERM signal.
+func handleSignal(sigCh chan os.Signal, vmConn net.Conn, proxyListener net.Listener) error {
+	if sigCh == nil {
+		return fmt.Errorf("Signal channel cannot be nil, it has to be initialized")
+	}
+
+	// Blocking here waiting for the signal to be received.
+	sig := <-sigCh
+	if sig != termSignal {
+		return fmt.Errorf("Signal received should be %q, got %q instead", termSignal.String(), sig.String())
+	}
+
+	// Let's first close the connection with the shim/runtime so that we
+	// don't get any more requests.
+	if proxyListener != nil {
+		if err := proxyListener.Close(); err != nil {
+			return err
+		}
+		proxyListener = nil
+	}
+
+	// Now let's close the connection with the VM/agent.
+	if vmConn != nil {
+		if err := vmConn.Close(); err != nil {
+			return err
+		}
+		vmConn = nil
```

Comment:
vm connection is always closed by main(). Besides, resetting vmConn here does not affect `servConn` in main() which means the connection will be closed again.

Reply:
That's why I test the value of `vmConn`, to prevent from multiple calls to Close().

Reply:
I meant there is no need to close `vmConn` in signal handler. It will always be closed in main's `defer` function.

And even if you close it here and set `vmConn` to nil, `servConn` in main is still non-nil and main will close the connection again.

Reply:
yeap, maybe `*net.Conn` will make a difference

Reply:
@bergwolf why do you think `vmConn` does not get updated from the main ?
`vmConn` being a `net.Conn`, and `net.Conn` being an interface related to object pointers, we can expect `vmConn` to be an actual pointer. If you set this pointer to nil, it will point to nil from the main too. Am I missing something ?

Reply:
And about the need to close `vmConn` from the signal handler, I agree this will be done eventually by the main, but I think it is better that we do that from the sig handler, the same way we close the proxy listener, leaving the `defer(){}` from the main being used only in case of error.

Reply:
@sboeuf `net.Conn` is an interface and is passed by pointer. See [the example ](https://play.golang.org/p/kuvE1d1d8yc) of it being the same even though being changed inside a function.

While I don't think it's a good practice to manipulate argument pointers like C here, I can live with `*net.Conn` if you insist.

Reply:
Yes you're right, I didn't realize it was not working as I was expecting...
Anyway, I have updated the PR so that I use pointer instead. I understand it's not the prettiest thing, but what I do like with this approach is that we make sure the disconnection with the shim append first, then the disconnection with the VM. The defers being there only to handle the error case.
If we use the defer to handle the normal case, this means we are going to wait for the results channel to be done, which happens when the connection with the VM is broken, which means the VM has been shutdown.
I expect that we can properly stop the proxy before the VM.